File: bin/mark_germlines.pl
#!/usr/bin/perl -w
use strict;
use Getopt::Long;
use File::Basename;
use lib dirname (__FILE__);
use vcf2;
use strict;
use JSON;
use Data::Dumper;

my $MAX_VAF_NORMAL = 0.05;
my $MIN_VAF_NORMAL = 0.35;
my $MIN_VAF_TUMOR  = 0.45;
my $MIN_DP         = 100;

# Get command line options
my %opt = ();
GetOptions( \%opt, 'vcf=s', 'tumor-id=s', 'normal-id=s' );
check_options( \%opt );

# Default to myeloid genes, for backwards compatibility
my %GENES;

my $vcf = vcf2->new('file'=>$opt{vcf} );

my $nid = ( $opt{'normal-id'} or 0 );
my $tid = ( $opt{'tumor-id'} or 0 );

#print_header($opt{vcf});

my %consequence_score = (
    "frameshift_variant"                => 11,
    "transcript_ablation"               => 10,
    "initiator_codon_variant"           => 9,
    "stop_gained"                       => 8,
    "start_lost"                        => 8,
    "stop_lost"                         => 8,
    "splice_acceptor_variant"           => 8,
    "splice_donor_variant"              => 8,
    "inframe_deletion"                  => 7,
    "transcript_amplification"          => 5,
    "splice_region_variant"             => 5,
    "missense_variant"                  => 5,
    "protein_altering_variant"          => 5,
    "inframe_insertion"                 => 5,
    "incomplete_terminal_codon_variant" => 5,
    "non_coding_transcript_exon_variant"=> 3,
    "synonymous_variant"                => 2,
    "mature_mirna_variant"              => 1,
    "non_coding_transcript_variant"     => 1,
    "regulatory_region_variant"         => 1,
    "upstream_gene_variant"             => 1,
    "regulatory_region_amplification"   => 1,
    "tfbs_amplification"                => 1,
    "5_prime_UTR_variant"               => 1,
    "intron_variant"                    => 1,
    "3_prime_UTR_variant"               => 1,
    "feature_truncation"                => 1,
    "TF_binding_site_variant"           => 1,
    "stop_retained_variant"             => 1,
    "feature_elongation"                => 1,
    "regulatory_region_ablation"        => 1,
    "tfbs_ablation"                     => 1,
    "coding_sequence_variant"           => 1,
    "downstream_gene_variant"           => 1,
    "NMD_transcript_variant"            => 1,
    "intergenic_variant"                => 0
);

my %clinvar_score = (
        "Benign" => -2,
        "Pathogenic" => 1,
        "Likely_pathogenic" => 1
);
my $inclusion_score    = 2;
my $gnomad_cutoff      = 0.01;
my $consequence_cutoff = 5;

while ( my $var = $vcf->next_var() ) {

    my $germline      = 0;
    my $germline_risk = 0;
    my $not_germline  = 0;
    
    # check whether it seems like a germline variant depending on VAF of variant
    for my $gt ( @{$var->{GT}}) {
        # If normal samples. Set GERMLINE filter if VAF > $MIN_VAF_NORMAL and in relevant gene
        if( $nid and $gt->{_sample_id} eq $nid ) {
            $germline = 1 if( $gt->{VAF} > $MIN_VAF_NORMAL and $gt->{DP} > $MIN_DP );
            $not_germline = 1 if $gt->{VAF} < $MAX_VAF_NORMAL and $gt->{DP} > $MIN_DP;
        }
        # Should we really do this? it just creates some false confidence that we can say something is germline in unpaired
        # If tumor samples. Set GERMLINE_RISK filter if VAF > $MIN_VAF_TUMOR
        # if( $tid and $gt->{_sample_id} eq $tid ) {
        #     $germline_risk = 1 if( $gt->{VAF} > $MIN_VAF_TUMOR and $gt->{DP} > $MIN_DP );
        # }
    }

    # try to limit GERMLINE variants
    my $score_results;
    if ($germline) {
        ($germline,$score_results) = mini_rank($var);
    }
    

    # Add GERMLINE filter and remove FAIL_NVAF for confirmed relevant germlines
    if ( $germline ) {
        my @new_filters = ("GERMLINE");
        foreach( split(';', $var->{FILTER}) ) {
            push @new_filters, $_ unless $_ eq "FAIL_NVAF"; 
        }
        $var->{FILTER} = join(";", @new_filters);
        #vcfstr($var);
        #print $var->{CHROM}.":".$var->{POS}
    }
    

    # Add GERMLINE_RISK filter to suspected germline variants in tumor only
    elsif ( $germline_risk and !$nid ) {
        my @filters = split ';', $var->{FILTER};
        push @filters, "GERMLINE_RISK";
        $var->{FILTER} = join(";", @filters);
    }

    #vcfstr($var);
}


sub print_header {
    my $file = shift;

    system("zgrep ^## $file");
    print "##FILTER=<ID=GERMLINE,Description=\"Germline variant, detected in normal sample\">\n";
    print "##FILTER=<ID=GERMLINE_RISK,Description=\"Potential germline variant, from tumor sample\">\n";
    system("zgrep ^#CHROM $file");
        
}


sub read_json {
    my $fn = shift;

    open( JSONFILE, $fn );
    my @json = <JSONFILE>;
    my $decoded = decode_json( join("", @json ) );
    close JSONFILE;

    return $decoded;
}

sub mini_rank {
    my $var = shift;
    my $score = 0;
    my %score_results;
    ## check clinvar ##
    # if only one annotation handle differently, this is to support sole Benign annotations
    # otherwise find highest score acoring to rank->clinvar from JSON. 
    # This let's the user define some value and let other be 0. This is ugly?
    my $clinvar = 0; my $clinvar_effect;
    if ($var->{INFO}->{CLNSIG}) {
        my @clinsig = split("/",$var->{INFO}->{CLNSIG});
        if (scalar(@clinsig) == 1) {
            if ($clinvar_score{$clinsig[0]}) {
                $clinvar = $clinvar_score{$clinsig[0]};
                $clinvar_effect = $clinsig[0];
            }
        }
        else {
            foreach my $match (@clinsig) {
                if ($clinvar_score{$match}) {
                    if ($clinvar_score{$match} > $clinvar) {
                        $clinvar = $clinvar_score{$match};
                        $clinvar_effect = $match;
                    }                    
                }
            }
        }
    }
    # add the clinvar score to score, Benign can disqualify the other two categories completely (is this ok?)
    if ($clinvar) {
        $score = $clinvar + $score;
        $score_results{"clinvar"}{"score"} = $score;
        $score_results{"clinvar"}{"effect"} = $clinvar_effect;
    }
    ## check consequence ## ## check gnomad ##
    # save the most severe consequence defined in JSON
    # check if gnomad passes cutoff from JSON
    my $max_score = 0; my $gnomad = 0; my $max_csq; my $gnomad_af = 0;
    for my $tx ( @{ $var->{INFO}->{CSQ} } ) {
        ($max_score,$max_csq) = conseqeunce($tx->{Consequence});
        if ( $tx->{gnomADg_AF} ) {
            my @afs = split(",",$tx->{gnomADg_AF});
            if ($afs[0]) {
                if ($afs[0] < $gnomad_cutoff) {
                    $gnomad = 1;
                    $gnomad_af = $afs[0];
                }
            }
        }
    }
    # add to score if most severe conseqeunce is within JSON cutoff

    if ($max_score >= $consequence_cutoff ) {
        $score++;
        $score_results{"csq"}{"score"} = $max_score;
        $score_results{"csq"}{"effect"} = $max_csq;
    }
    # add to score if gnomad is below cutoff
    if ($gnomad > 0) {
        $score++;
        $score_results{"gnomad"}{"score"} = $gnomad;
        $score_results{"gnomad"}{"af"} = $gnomad_af;
    }
    # if score higher than or equal to inclusion score return true and retain GERMLINE status
    if ($score >= $inclusion_score ) {
        print Dumper(%score_results);
        return 1,\%score_results;
    }
    # else send back 0, and thus GERMLINE is no longer true
    else {
        return 0,\%score_results;
    }
}

sub conseqeunce {
    my $tx = shift;
    my $max_score = 0;
    my $max_effect;
    foreach my $csq (@{$tx} ) {
        if ($consequence_score{$csq}) {
            if ( $consequence_score{$csq} > $max_score ) {
                $max_score = $consequence_score{$csq};
                $max_effect = $csq;
            }
        }
    }
    return $max_score,$max_effect;
}

sub check_options {
    my %opt = %{ $_[0] };

    help_text() unless $opt{vcf};

    die "File does not exist $opt{vcf}..." if ! -s $opt{vcf};

}


sub help_text {
    my $error = shift;
    
    print "\n\$ mark_germlines.pl --vcf INPUT_VCF --tumor-id [--normal-id}\n\n";
    print "   --vcf        Input vcf\n";
    print "   --normal-id  Normal sample ID\n";
    print "   --tumor-id   Tumor sample ID\n";
    print "   --assay      Select an assay. Determines which genes are included\n";
    print "   --assay_json Select an assay according to input json, filter uninteresing GERMLINE for select genes";
    print "\n";
    exit(0);
}


sub vcfstr {
    my( $v ) = @_;

    my @all_info;
    print $v->{CHROM}."\t".$v->{POS}."\t".$v->{ID}."\t".$v->{REF}."\t".$v->{ALT}."\t".$v->{QUAL}."\t".$v->{FILTER}."\t";

    # Generate and print INFO field
    for my $info_key (@{$v->{INFO_order}}) {
    my $key2 = $info_key;
    $key2 = "_CSQ_str" if $info_key eq "CSQ";
    push @all_info, $info_key."=".$v->{INFO}->{$key2};
    }
    print join(";", @all_info)."\t";

    # Print FORMAT field
    print join(":", @{$v->{FORMAT}});

    
    # Print GT fields for all samples
    for my $gt (@{$v->{GT}}) {
    my @all_gt;
    for my $key ( @{$v->{FORMAT}} ) {
        push @all_gt, ( defined $gt->{$key} ? $gt->{$key} : "");
    }
    print "\t".join(":", @all_gt);
    }
    print "\n";
}

File: bin/mark_germlines.py
"""Germline-related VCF utilities."""

from __future__ import annotations

import json
from argparse import ArgumentParser

import vcf2 as vcf
from vcf_pipeline_utils import add_info, leading_float, open_output, write_filter_header, write_variant


def max_consequence_score(consequences, rank):
    """Return the highest consequence score from an assay definition."""
    scores = rank.get("consequence_score", {})
    return max((scores.get(item, 0) for item in consequences), default=0)


def passes_assay_rank(var, assay):
    """Return whether a germline candidate passes assay inclusion scoring."""
    score = 0
    clinvar_score = 0
    if var["INFO"].get("CLNSIG") and assay.get("clinvar"):
        for value in str(var["INFO"]["CLNSIG"]).split("/"):
            clinvar_score = max(clinvar_score, assay["clinvar"].get(value, 0))
    score += clinvar_score

    max_score = 0
    gnomad = 0
    for tx in var["INFO"].get("CSQ", []):
        if tx.get("Consequence") and assay.get("consequence_cutoff") and assay.get("consequence_score"):
            max_score = max(max_score, max_consequence_score(tx["Consequence"], assay))
        if (tx.get("gnomADg_AF") or tx.get("gnomAD_AF")) and assay.get("gnomad_cutoff"):
            genome = leading_float(str(tx.get("gnomADg_AF", "")).split(",")[0])
            exome = leading_float(str(tx.get("gnomAD_AF", "")).split(",")[0])
            if (genome or exome) < leading_float(assay["gnomad_cutoff"]):
                gnomad = 1

    if assay.get("consequence_cutoff") and max_score >= leading_float(assay["consequence_cutoff"]):
        score += 1
    if gnomad >= 1:
        score += 1
    return score >= leading_float(assay.get("inclusion_score"))


def mark_germlines(
    vcf_file,
    tumor_id,
    normal_id,
    assay_file,
    out_file,
    min_normal_vaf=0.35,
    min_tumor_vaf=0.45,
    min_depth=100,
):
    """Add GERMLINE and GERMLINE_RISK filters to a VCF."""
    with open(assay_file) as fh:
        assay = json.load(fh)
    genes = {gene: True for gene in assay.get("genes", [])}

    headers = [
        '##FILTER=<ID=GERMLINE,Description="Germline variant, detected in normal sample">',
        '##FILTER=<ID=GERMLINE_RISK,Description="Potential germline variant, from tumor sample">',
    ]
    with vcf.VCFReader(vcf_file) as reader, open_output(out_file) as out_fh:
        write_filter_header(reader, out_fh, headers)
        for var in reader:
            in_gene = any(genes.get(tx.get("SYMBOL")) or genes.get("ALL_GENES") for tx in var["INFO"].get("CSQ", []))
            germline = False
            germline_risk = False
            for gt in var["GT"]:
                if normal_id and gt["_sample_id"] == normal_id:
                    germline = (
                        leading_float(gt.get("VAF")) > min_normal_vaf
                        and leading_float(gt.get("DP")) > min_depth
                        and in_gene
                    )
                if tumor_id and gt["_sample_id"] == tumor_id:
                    germline_risk = (
                        leading_float(gt.get("VAF")) > min_tumor_vaf and leading_float(gt.get("DP")) > min_depth
                    )

            if assay.get("inclusion_score") and germline:
                germline = passes_assay_rank(var, assay)

            if germline:
                filters = ["GERMLINE"] + [item for item in var["FILTER"].split(";") if item != "FAIL_NVAF"]
                var["FILTER"] = ";".join(filters)
            elif germline_risk and not normal_id:
                var["FILTER"] = ";".join(var["FILTER"].split(";") + ["GERMLINE_RISK"])
            write_variant(var, out_fh, use_original_csq=True, trailing_sample_tab=False)


def germline_for_cnvkit(vcf_file, out_file, min_gnomad_af=0.05, min_variant_depth=50):
    """Keep common germline variants suitable for CNVkit BAF input."""
    with vcf.VCFReader(vcf_file) as reader, open_output(out_file) as out_fh:
        for line in reader.meta_header_lines:
            if line.startswith("##INFO") and "AF,Number=A" not in line:
                print(line, file=out_fh)
                print(
                    '##INFO=<ID=AF,Number=A,Type=Float,Description="Estimated allele frequency in the range (0,1]">',
                    file=out_fh,
                )
            elif "ID=VAF" not in line:
                print(line, file=out_fh)
        print(reader.column_header_line, file=out_fh)
        for var in reader:
            vaf_value = ""
            vd = 0
            for gt in var["GT"]:
                vaf_value = gt.get("VAF", "")
                vd = leading_float(gt.get("VD"))
            gnomad = 0
            if var["INFO"].get("CSQ") and var["INFO"]["CSQ"][0].get("gnomADg_AF"):
                gnomad = leading_float(var["INFO"]["CSQ"][0].get("gnomADg_AF"))
            if not (gnomad >= min_gnomad_af and vd >= min_variant_depth):
                continue
            add_info(var, "AF", vaf_value)
            write_variant(var, out_fh, use_original_csq=True)


def build_mark_parser():
    """Build the mark-germlines parser."""
    parser = ArgumentParser(description="Mark likely germline variants in a VCF.")
    parser.add_argument("--vcf", required=True, help="Input VCF file.")
    parser.add_argument("--tumor-id", help="Tumor sample ID.")
    parser.add_argument("--normal-id", help="Normal sample ID.")
    parser.add_argument("--assay", required=True, help="Assay JSON file.")
    parser.add_argument("--out", required=True, help="Output VCF file.")
    parser.add_argument("--min-normal-vaf", type=float, default=0.35)
    parser.add_argument("--min-tumor-vaf", type=float, default=0.45)
    parser.add_argument("--min-depth", type=float, default=100)
    return parser


def main_mark(argv=None):
    """Run mark-germlines CLI."""
    args = build_mark_parser().parse_args(argv)
    mark_germlines(
        args.vcf,
        args.tumor_id,
        args.normal_id,
        args.assay,
        args.out,
        args.min_normal_vaf,
        args.min_tumor_vaf,
        args.min_depth,
    )
    return 0

if __name__ == "__main__":
    raise SystemExit(main_mark())
